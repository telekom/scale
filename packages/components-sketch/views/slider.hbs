{{!--
Scale https://github.com/telekom/scale

Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at https://mozilla.org/MPL/2.0/.
--}}

<div
  data-sketch-artboard="Slider Track Small"
  class="artboard"
  data-category="Small"
>
  <h1 class="title--artboard">Slider Track Small</h1>

  <h2 class="title--artboard-section">Label On - Counter On</h2>
  <div class="agent-states--vertical-grid">
    <div class="agent-states--item">
      <scale-slider
        data-sketch-key="slider-1"
        label="Slider Label"
        value="40"
        track-small
      ></scale-slider>
    </div>
    <div class="agent-states--item">magenta</div>
    <div class="agent-states--item">
      <scale-slider
        data-sketch-key="slider-2"
        label="Slider Label"
        value="40"
        track-small
        custom-color="#009DE0"
      >
      </scale-slider>
    </div>
    <div class="agent-states--item">custom color</div>
    <div class="agent-states--item">
      <scale-slider
        data-sketch-key="slider-3"
        label="Slider Label"
        value="40"
        track-small
        disabled
      ></scale-slider>
    </div>
    <div class="agent-states--item">disabled</div>
  </div>
</div>

<div
  data-sketch-artboard="Slider Track Small - Large Thumb"
  class="artboard"
  data-category="Small - Large Thumb"
>
  <h1 class="title--artboard">Slider Track Small - Large Thumb</h1>

  <h2 class="title--artboard-section">Label On - Counter On</h2>
  <div class="agent-states--vertical-grid">
    <div class="agent-states--item">
      <scale-slider
        data-sketch-key="slider-s-lt-standard"
        thumb-large="true"
        label="Slider Label"
        value="40"
        track-small
      ></scale-slider>
    </div>
    <div class="agent-states--item">magenta</div>
    <div class="agent-states--item">
      <scale-slider
        data-sketch-key="slider-s-lt-custom"
        thumb-large="true"
        label="Slider Label"
        value="40"
        track-small
        custom-color="#009DE0"
      ></scale-slider>
    </div>
    <div class="agent-states--item">custom color</div>
    <div class="agent-states--item">
      <scale-slider
        data-sketch-key="slider-s-lt-disabled"
        thumb-large="true"
        label="Slider Label"
        value="40"
        track-small
        disabled
      ></scale-slider>
    </div>
    <div class="agent-states--item">disabled</div>
  </div>
</div>

<div
  data-sketch-artboard="Slider Track Large"
  class="artboard"
  data-category="Large"
>
  <h1 class="title--artboard">Slider Track Large</h1>

  <h2 class="title--artboard-section">Label On - Counter On</h2>
  <div class="agent-states--vertical-grid">
    <div class="agent-states--item">
      <scale-slider
        data-sketch-key="slider-13"
        label="Slider Label"
        value="40"
        track-height="5"
      ></scale-slider>
    </div>
    <div class="agent-states--item">magenta</div>
    <div class="agent-states--item">
      <scale-slider
        data-sketch-key="slider-14"
        label="Slider Label"
        value="40"
        track-height="5"
        custom-color="#009DE0"
      >
      </scale-slider>
    </div>
    <div class="agent-states--item">custom color</div>
    <div class="agent-states--item">
      <scale-slider
        data-sketch-key="slider-15"
        label="Slider Label"
        value="40"
        track-height="5"
        disabled
      >
      </scale-slider>
    </div>
    <div class="agent-states--item">disabled</div>
  </div>
</div>

<div
  data-sketch-artboard="Slider Track Large - Large Thumb"
  class="artboard"
  data-category="Large Thumb"
>
  <h1 class="title--artboard">Slider Track Large - Large Thumb</h1>

  <h2 class="title--artboard-section">Label On - Counter On</h2>
  <div class="agent-states--vertical-grid">
    <div class="agent-states--item">
      <scale-slider
        data-sketch-key="slider-lt-13"
        thumb-large="true"
        label="Slider Label"
        value="40"
        track-height="5"
      >
      </scale-slider>
    </div>
    <div class="agent-states--item">magenta</div>
    <div class="agent-states--item">
      <scale-slider
        data-sketch-key="slider-lt-14"
        thumb-large="true"
        label="Slider Label"
        value="40"
        track-height="5"
        custom-color="#009DE0"
      ></scale-slider>
    </div>
    <div class="agent-states--item">custom color</div>
    <div class="agent-states--item">
      <scale-slider
        data-sketch-key="slider-lt-15"
        thumb-large="true"
        label="Slider Label"
        value="40"
        track-height="5"
        disabled
      ></scale-slider>
    </div>
    <div class="agent-states--item">disabled</div>
  </div>
</div>

<script src="/symbol_names.js"></script>
<script>
  Array.from(document.querySelectorAll('scale-slider')).forEach((btn) => {
    btn.dataset.sketchSymbol = `Slider / ${order(
      0,
      getCategoryName(btn)
    )} / ${order(
      4,
      btn.hasAttribute('disabled')
        ? 'Disabled'
        : btn.hasAttribute('custom-color')
        ? 'Custom Color'
        : 'Magenta'
    )}`;
  });
</script>
