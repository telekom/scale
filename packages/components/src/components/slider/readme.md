# t-slider



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute      | Description                         | Type      | Default |
| ------------- | -------------- | ----------------------------------- | --------- | ------- |
| `customClass` | `custom-class` | (optional) Card class               | `string`  | `''`    |
| `label`       | `label`        | (optional) Slider display value     | `boolean` | `false` |
| `max`         | `max`          | (optional) Slider range max value   | `number`  | `100`   |
| `min`         | `min`          | (optional) Slider range start value | `number`  | `0`     |
| `step`        | `step`         | (optional) Slider step              | `number`  | `1`     |
| `value`       | `value`        | (optional) Slider binding value     | `number`  | `0`     |
| `vertical`    | `vertical`     | (optional) Slider direction         | `boolean` | `false` |


## CSS Custom Properties

| Name                                 | Description                                                                                |
| ------------------------------------ | ------------------------------------------------------------------------------------------ |
| `--slider-input-background`          | Background of the slider input                                                             |
| `--slider-input-border-radius`       | Border radius of the slider input default                                                  |
| `--slider-input-height`              | Height of the slider input                                                                 |
| `--slider-input-thumb-background`    | Background of the slider input thumb @Prop --slider-text-margin: Margin of the slider text |
| `--slider-input-thumb-border`        | Border of the slider input default thumb                                                   |
| `--slider-input-thumb-border-radius` | Border radius of the slider input default thumb                                            |
| `--slider-input-thumb-height`        | Height of the slider input default thumb                                                   |
| `--slider-input-thumb-width`         | Width of the slider input default thumb                                                    |
| `--slider-input-width`               | Width of the slider input                                                                  |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
