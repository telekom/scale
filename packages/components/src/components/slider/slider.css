/**
* @prop --slider-input-background: Background of the slider input
* @prop --slider-input-height: Height of the slider input
* @prop --slider-input-width: Width of the slider input
* @prop --slider-input-border-radius: Border radius of the slider input default
* @prop --slider-input-thumb-border: Border of the slider input default thumb
* @prop --slider-input-thumb-height: Height of the slider input default thumb
* @prop --slider-input-thumb-width: Width of the slider input default thumb
* @prop --slider-input-thumb-border-radius: Border radius of the slider input default thumb
* @prop --slider-input-thumb-background: Background of the slider input thumb
* @Prop --slider-text-margin: Margin of the slider text
*/

.slider {
  display: inline-flex;
  align-items: center;
  width: 100%;
}

.slider__input {
  outline: none;
  -webkit-appearance: none;
  cursor: pointer;
  background: var(--slider-input-background, #e4e7ed);
  height: var(--slider-input-height, 4px);
  width: var(--slider-input-width, 100%);
  border-radius: var(--slider-input-border-radius, 2px);
}

.slider__input::-webkit-slider-thumb {
  -webkit-appearance: none;
  border: var(--slider-input-thumb-border, 2px solid #409eff);
  height: var(--slider-input-thumb-height, 20px);
  width: var(--slider-input-thumb-width, 20px);
  border-radius: var(--slider-input-thumb-border-radius, 50%);
  background: var(--slider-input-thumb-background, #fff);
}

.slider__input::-moz-range-thumb {
  -webkit-appearance: none;
  border: var(--slider-input-thumb-border, 2px solid #409eff);
  height: var(--slider-input-thumb-height, 20px);
  width: var(--slider-input-thumb-width, 20px);
  border-radius: var(--slider-input-thumb-border-radius, 50%);
  background: var(--slider-input-thumb-background, #fff);
}

.slider__input::-ms-thumb {
  -webkit-appearance: none;
  border: var(--slider-input-thumb-border, 2px solid #409eff);
  height: var(--slider-input-thumb-height, 20px);
  width: var(--slider-input-thumb-width, 20px);
  border-radius: var(--slider-input-thumb-border-radius, 50%);
  background: var(--slider-input-thumb-background, #fff);
}

input[type='range']::-moz-focus-outer {
  border: 0;
}

.slider__text {
  margin: var(--slider-text-margin, 0 0 0 1rem);
  min-width: 30px;
  text-align: right;
}
