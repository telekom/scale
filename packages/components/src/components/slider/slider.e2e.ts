import { newE2EPage } from '@stencil/core/testing';

describe('t-slider', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<t-slider/>');
    const element = await page.find('t-slider');
    expect(element).toHaveClass('hydrated');
  });
});
