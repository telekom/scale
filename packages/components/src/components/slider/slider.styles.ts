import { JssStyle } from 'jss';

const sliderKnob: JssStyle = {
  '-webkit-appearance': 'none',
  border: '2px solid #409eff',
  height: '20px',
  width: '20px',
  borderRadius: '50%',
  background: '#fff',
};

export const styles: JssStyle = {
  slider: {
    display: 'inline-flex',
    alignItems: 'center',
    width: '100%',
  },
  slider__text: {
    margin: 'var(--slider-text-margin, 0 0 0 1rem)',
    minWidth: 30,
    textAlign: 'right',
  },
  slider__input: {
    outline: 'none',
    '-webkit-appearance': 'none',
    cursor: 'pointer',
    background: '#e4e7ed',
    height: '4px',
    width: '100%',
    borderRadius: '2px',
    '&::-webkit-slider-thumb': sliderKnob,

    // TODO: currently throws warning in chrome, enable browser check to support firefox and edge
    // '&::-moz-range-thumb': sliderKnob,
    // '&::-ms-thumb': sliderKnob,
  },
  // TODO: currently throws warning in chrome, enable browser check to support firefox
  // "input[type='range']": {
  //   '&::-moz-focus-outer': {
  //     border: '0',
  //   },
  // },
};
