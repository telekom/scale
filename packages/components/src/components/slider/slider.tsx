import { Component, h, Prop } from '@stencil/core';
import { CssClassMap } from '../../utils/utils';
import classNames from 'classnames';

@Component({
  tag: 't-slider',
  styleUrl: 'slider.css',
})
export class Slider {
  /** (optional) Card class */
  @Prop() public customClass?: string = '';
  /** (optional) Slider direction */
  @Prop() public vertical?: boolean = false;
  /** (optional) Slider range start value */
  @Prop() public min?: number = 0;
  /** (optional) Slider range max value */
  @Prop() public max?: number = 100;
  /** (optional) Slider binding value */
  @Prop() public value: number = 0;
  /** (optional) Slider step */
  @Prop() public step: number = 1;
  /** (optional) Slider display value */
  @Prop() public label: boolean = false;

  public componentWillLoad() {
    if (this.min > this.value) {
      this.value = this.min;
    }
  }

  public render() {
    return (
      <div class={this.getCssClassMap()}>
        <input
          class="slider__input"
          type="range"
          min={this.min}
          max={this.max}
          value={this.value}
          step={this.step}
          onInput={this.updateValue}
        />
        {this.label && <span class="slider__text">{this.value}</span>}
      </div>
    );
  }

  private updateValue = event => {
    this.value = event.target.value;
  };

  private getCssClassMap(): CssClassMap {
    return classNames('slider', this.customClass && this.customClass);
  }
}
