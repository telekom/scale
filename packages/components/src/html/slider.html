<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" /
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Slider</title>

    <script type="module" src="/build/scale-components.esm.js"></script>
    <link rel="stylesheet" href="/build/scale-components.css" />
    <style>
      demo-hue {
        display: flex;
      }
      demo-color-display {
        display: block;
        width: 120px;
        height: 120px;
        border-radius: 60px;
        background: hsla(var(--h) var(--s) var(--l) / var(--a));
        margin-right: 2rem;
        box-shadow: var(--scl-shadow-level-2);
      }
    </style>
  </head>
  <body>
    <h1 class="scl-font-variant-heading-3">Color thingy</h1>
    <demo-hue>
      <demo-color-display
        style="--h: 110deg; --s: 100%; --l: 44.3%; --a: 1"
      ></demo-color-display>
      <demo-controls>
        <scale-slider
          id="h"
          label="Hue"
          value="110"
          max="360"
          step="0.1"
          decimals="1"
          unit="deg"
        ></scale-slider>
        <scale-slider id="s" label="Saturation" value="100"></scale-slider>
        <scale-slider
          id="l"
          label="Lightness"
          value="44.3"
          step="0.1"
          decimals="1"
        ></scale-slider>
        <scale-slider
          id="a"
          label="Alpha"
          unit=""
          decimals="2"
          value="1"
          min="0"
          max="1"
          step="0.01"
        ></scale-slider>
      </demo-controls>
    </demo-hue>

    <script>
      const h = document.getElementById('h');
      const s = document.getElementById('s');
      const l = document.getElementById('l');
      const a = document.getElementById('a');
      const display = document.querySelector('demo-color-display');

      h.addEventListener('scaleInput', () => {
        display.style.setProperty('--h', h.value + 'deg');
      });
      s.addEventListener('scaleInput', () => {
        display.style.setProperty('--s', s.value + '%');
      });
      l.addEventListener('scaleInput', () => {
        display.style.setProperty('--l', l.value + '%');
      });
      a.addEventListener('scaleInput', () => {
        display.style.setProperty('--a', a.value);
      });
    </script>
  </body>
</html>
