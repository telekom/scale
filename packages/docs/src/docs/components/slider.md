---
title: Slider
description: Slider element
url: /docs/slider
contributors:
  - maomaoZH
  - eeegor
---

# Slider

Slider component

selector: `t-slider`

## Slider

```javascript
<t-slider></t-slider>
```

<div class="demo-container">
  <t-slider></t-slider>
</div>

## With label

```javascript
<t-slider label></t-slider>
```

<div class="demo-container">
  <t-slider label></t-slider>
</div>

## Step 10

```javascript
<t-slider label step="10"></t-slider>
```

<div class="demo-container">
  <t-slider label step="10"></t-slider>
</div>

## With min and max value

```javascript
<t-slider label min="10" max="1000"></t-slider>
```

<div class="demo-container">
  <t-slider label min="10" max="1000"></t-slider>
</div>
