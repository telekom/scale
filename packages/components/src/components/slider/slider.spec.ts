import { newSpecPage } from '@stencil/core/testing';
import { Slider } from './slider';

describe('Slider', () => {
  let element;
  beforeEach(async () => {
    element = new Slider();
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Slider],
      html: `<t-slider>Label</t-slider>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should update value when slider moved', () => {
    const event = {
      target: {
        value: 30,
        style: {
          background:
            'linear-gradient(to right, #409eff 0%, #409eff 0%, #e4e7ed 0%, #e4e7ed 100%)',
        },
      },
    };
    element.value = 0;
    element.updateValue(event);
    expect(element.value).toBe(30);
  });
});
