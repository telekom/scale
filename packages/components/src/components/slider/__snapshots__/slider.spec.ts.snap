// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider should match snapshot 1`] = `
<scale-slider>
  <mock:shadow-root>
    <style>
      .slider-0-0-4 {
  width: 100%;
  display: inline-flex;
  align-items: center;
}
.slider__text-0-0-5 {
  margin: var(--slider-text-margin, 0 0 0 1rem);
  min-width: 30px;
  text-align: right;
}
.slider__input-0-0-6 {
  width: 100%;
  cursor: pointer;
  height: 4px;
  outline: none;
  background: #e4e7ed;
  border-radius: 2px;
  -webkit-appearance: none;
}
.slider__input-0-0-6::-webkit-slider-thumb {
  width: 20px;
  border: 2px solid #409eff;
  height: 20px;
  background: #fff;
  border-radius: 50%;
  -webkit-appearance: none;
}
    </style>
    <div class="slider-0-0-4">
      <input class="slider__input-0-0-6" max="100" min="0" step="1" type="range" value="0">
    </div>
  </mock:shadow-root>
  Label
</scale-slider>
`;
